File: RestfulObjects Server/RestfulObjects.Test.Data/WithValue.cs
// Copyright © Naked Objects Group Ltd ( http://www.nakedobjects.net). 
// All Rights Reserved. This code released under the terms of the 
// Microsoft Public License (MS-PL) ( http://opensource.org/licenses/ms-pl.html) 

using System;
using System.ComponentModel.DataAnnotations;
using NakedObjects;
using NakedObjects.Security;

namespace RestfulObjects.Test.Data {

    [PresentationHint("class1 class2")]
    public class WithValue {
        private DateTime aDateTimeValue = new DateTime(2012, 2, 10);

        [Key, Title, ConcurrencyCheck]
        public virtual int Id { get; set; }

        [PresentationHint("class3 class4")]
        public virtual int AValue { get; set; }

        [Disabled]
        public virtual int ADisabledValue { get; set; }

        [Hidden]
        public virtual int AHiddenValue { get; set; }

        public virtual int AChoicesValue { get; set; }

        [System.ComponentModel.DataAnnotations.MaxLength(101)]
        [RegEx(Validation = @"[A-Z]")]
        [Optionally]
        [DescribedAs("A string value for testing")]
        [MemberOrder(Sequence = "3")]
        public virtual string AStringValue { get; set; }

        [Optionally]
        [DescribedAs("A datetime value for testing")]
        [Mask("d")]
        [MemberOrder(Sequence = "4")]
        public virtual DateTime ADateTimeValue {
            get { return aDateTimeValue; }
            set { aDateTimeValue = value; }
        }

        [AuthorizeProperty(ViewUsers = "viewUser")]
        public virtual int AUserHiddenValue { get; set; }

        [AuthorizeProperty(EditUsers = "editUser")]
        public virtual int AUserDisabledValue { get; set; }

        public virtual int[] ChoicesAChoicesValue() {
            return new[] {1, 2, 3};
        }

        public virtual string Validate(int aValue, int aChoicesValue) {
            if (aValue == 101 && aChoicesValue == 3) {
                return "Cross validation failed";
            }
            return "";
        }

        public virtual int AConditionalChoicesValue { get; set; }

        public virtual int[] ChoicesAConditionalChoicesValue(int aValue, string aStringValue) {
            return new[] {  aValue, aStringValue == null ? 0 : int.Parse(aStringValue)  };
        }
    }
}

File: Rest/NakedObjects.Rest.Test.Data.Sdm2/WithValue.vb
' Copyright © Naked Objects Group Ltd ( http://www.nakedobjects.net). 
' All Rights Reserved. This code released under the terms of the 
' Microsoft Public License (MS-PL) ( http://opensource.org/licenses/ms-pl.html) 
Imports sdm.systems.application.control

Namespace RestfulObjects.Test.Data
    Public Class WithValue
        Inherits AbstractBusinessObject

#Region "Id and Title"

        Public Overrides Function title() As Title
            Dim t As New Title
            t.append(Id)
            Return t
        End Function

        Private _id As New WholeNumber

        Public ReadOnly Property Id() As WholeNumber
            Get
                Return _id
            End Get
        End Property

#End Region

#Region "AValue"

        Private _aValue As New WholeNumber

        Public Property AValue() As WholeNumber
            Get
                Return _aValue
            End Get
            Set(value As WholeNumber)
                _aValue = value
            End Set
        End Property

#End Region

#Region "AStringValue"

        Private _aStringValue As New TextString

        Public Property AStringValue() As TextString
            Get
                Return _aStringValue
            End Get
            Set(value As TextString)
                _aStringValue = value
            End Set
        End Property


#End Region

#Region "ADateTimeValue"

        Private _aDateTimeValue As New NoDate(Me, 2012, 2, 10)

        Public Property ADateTimeValue() As NoDate
            Get
                Return _aDateTimeValue
            End Get
            Set(value As NoDate)
                _aDateTimeValue = value
            End Set
        End Property


#End Region


#Region "ADisabledValue"

        Private _aDisabledValue As New WholeNumber

        Public ReadOnly Property ADisabledValue() As WholeNumber
            Get
                Return _aDisabledValue
            End Get
        End Property

        Public Overridable Sub aboutADisabledValue(ByVal a As FieldAbout, ByVal value As WholeNumber)
            a.unmodifiable("Field not editable")
        End Sub

#End Region

#Region "AHiddenValue"

        Private _aHiddenValue As New WholeNumber

        Public ReadOnly Property AHiddenValue() As WholeNumber
            Get
                Return _aHiddenValue
            End Get
        End Property

        Public Overridable Sub aboutAHiddenValue(ByVal a As FieldAbout, ByVal value As WholeNumber)
            a.invisible()
        End Sub

#End Region

#Region "AChoicesValue"

        Private _aChoicesValue As New WholeNumber

        Public ReadOnly Property AChoicesValue() As WholeNumber
            Get
                Return _aChoicesValue
            End Get
        End Property

        Public Overridable Sub aboutAChoicesValue(ByVal a As FieldAbout, ByVal value As WholeNumber)
            a.cannotBeEmpty(value)
        End Sub

        'RP:  Choices exist for parameters, not for properties.
        'Public Overridable Function ChoicesAChoicesValue() As Integer()
        '    Return {1, 2, 3}
        'End Function

#End Region

#Region "Set up (Fixtures)"

        Public Shared Function SetUp(id As Integer, val As Integer, dis As Integer) As WithValue
            Dim wv As WithValue =
                    CType(
                        createTransientInstanceFromShared(
                            GetType(WithValue)),
                        WithValue)


            wv.Id.setValue(id)
            wv.AValue.setValue(val)
            wv.ADisabledValue.setValue(dis)
            wv.actionSave()
            Return wv
        End Function

#End Region
    End Class
End Namespace